File: kotlin/src/test/kotlin/aether/security/SignalFixtureTest.kt
// SPDX-License-Identifier: MIT
package aether.security

import org.bouncycastle.crypto.agreement.X25519Agreement
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import org.bouncycastle.crypto.params.X25519PublicKeyParameters
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.io.File
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Cross-language Signal-protocol fixture verifier and end-to-end exercises.
 *
 * Verifies that the Kotlin implementation produces byte-identical X3DH and
 * ratchet outputs to the C# reference (committed in
 * fixtures/signal/expected/*.json). Any drift between Kotlin and the other
 * 7 languages surfaces here as a hex mismatch.
 */
class SignalFixtureTest {

    // ─── Fixture verifiers ──────────────────────────────────────────────────

    @Test
    fun `signal fixture x3dh_basic`() {
        val (inputs, expected) = loadFixturePair("x3dh_basic")

        val aliceIK = unhex(inputs["alice_identity_priv_hex"]!!)
        val aliceEK = unhex(inputs["alice_ephemeral_priv_hex"]!!)
        val bobIK = unhex(inputs["bob_identity_priv_hex"]!!)
        val bobSPK = unhex(inputs["bob_signed_pre_key_priv_hex"]!!)
        val bobOPK = unhex(inputs["bob_one_time_pre_key_priv_hex"]!!)

        val aliceIKPub = x25519DerivePub(aliceIK)
        val aliceEKPub = x25519DerivePub(aliceEK)
        val bobIKPub = x25519DerivePub(bobIK)
        val bobSPKPub = x25519DerivePub(bobSPK)
        val bobOPKPub = x25519DerivePub(bobOPK)

        val dh1 = x25519AgreeStandalone(aliceIK, bobSPKPub)
        val dh2 = x25519AgreeStandalone(aliceEK, bobIKPub)
        val dh3 = x25519AgreeStandalone(aliceEK, bobSPKPub)
        val dh4 = x25519AgreeStandalone(aliceEK, bobOPKPub)

        val shared = dh1 + dh2 + dh3 + dh4
        val rootInfo = inputs["hkdf_root_info_utf8"]!!.toByteArray(Charsets.UTF_8)
        val sendInfo = inputs["hkdf_chain_initiator_send_info_utf8"]!!.toByteArray(Charsets.UTF_8)
        val recvInfo = inputs["hkdf_chain_initiator_recv_info_utf8"]!!.toByteArray(Charsets.UTF_8)

        val rootKey = hkdf32Standalone(shared, rootInfo)
        val sendChain = hkdf32Standalone(rootKey, sendInfo)
        val recvChain = hkdf32Standalone(rootKey, recvInfo)

        assertEquals(expected["alice_identity_pub_hex"], hex(aliceIKPub))
        assertEquals(expected["alice_ephemeral_pub_hex"], hex(aliceEKPub))
        assertEquals(expected["bob_identity_pub_hex"], hex(bobIKPub))
        assertEquals(expected["bob_signed_pre_key_pub_hex"], hex(bobSPKPub))
        assertEquals(expected["bob_one_time_pre_key_pub_hex"], hex(bobOPKPub))
        assertEquals(expected["dh1_hex"], hex(dh1))
        assertEquals(expected["dh2_hex"], hex(dh2))
        assertEquals(expected["dh3_hex"], hex(dh3))
        assertEquals(expected["dh4_hex"], hex(dh4))
        assertEquals(expected["shared_secret_hex"], hex(shared))
        assertEquals(expected["root_key_hex"], hex(rootKey))
        assertEquals(expected["initiator_send_chain_key_hex"], hex(sendChain))
        assertEquals(expected["initiator_recv_chain_key_hex"], hex(recvChain))
    }

    @Test
    fun `signal fixture ratchet_step_basic`() {
        val (inputs, expected) = loadFixturePair("ratchet_step_basic")
        val chainKey = unhex(inputs["chain_key_hex"]!!)
        assertEquals(expected["message_key_hex"], hex(hmacOne(chainKey, 0x01)))
        assertEquals(expected["next_chain_key_hex"], hex(hmacOne(chainKey, 0x02)))
    }

    @Test
    fun `signal fixture ratchet_step_three_iterations`() {
        val (inputs, expected) = loadFixturePair("ratchet_step_three_iterations")
        var chainKey = unhex(inputs["initial_chain_key_hex"]!!)
        for (i in 0 until 3) {
            val msg = hmacOne(chainKey, 0x01)
            val nxt = hmacOne(chainKey, 0x02)
            assertEquals(expected["step_${i}_message_key_hex"], hex(msg))
            assertEquals(expected["step_${i}_chain_key_after_hex"], hex(nxt))
            chainKey = nxt
        }
    }

    // ─── End-to-end exercises ──────────────────────────────────────────────

    @Test
    fun `X3DH first message round-trips`() {
        val alice = SignalProtocol()
        val bob = SignalProtocol()
        val bobBundle = bob.generatePreKeyBundle("bob")
        alice.generatePreKeyBundle("alice")
        alice.processPreKeyBundle(bobBundle)

        val encrypted = alice.encrypt("bob", "the mesh is alive".toByteArray())
        assertEquals(SignalProtocol.MESSAGE_TYPE_PRE_KEY, encrypted.messageType)
        assertEquals(32, encrypted.initiatorIdentityKeyX25519?.size)
        assertEquals(32, encrypted.initiatorEphemeralKeyX25519?.size)
        assertEquals("alice", encrypted.senderUhid)

        val plaintext = bob.decrypt("alice", encrypted)
        assertEquals("the mesh is alive", String(plaintext))
        assertTrue(bob.hasSession("alice"))
    }

    @Test
    fun `X3DH subsequent message is normal not pre-key`() {
        val alice = SignalProtocol()
        val bob = SignalProtocol()
        val bobBundle = bob.generatePreKeyBundle("bob")
        alice.generatePreKeyBundle("alice")
        alice.processPreKeyBundle(bobBundle)

        val first = alice.encrypt("bob", "a".toByteArray())
        bob.decrypt("alice", first)

        val second = alice.encrypt("bob", "b".toByteArray())
        assertEquals(SignalProtocol.MESSAGE_TYPE_NORMAL, second.messageType)
        assertNull(second.initiatorIdentityKeyX25519)

        val out = bob.decrypt("alice", second)
        assertEquals("b", String(out))
    }

    @Test
    fun `X3DH bidirectional after first message`() {
        val alice = SignalProtocol()
        val bob = SignalProtocol()
        val bobBundle = bob.generatePreKeyBundle("bob")
        alice.generatePreKeyBundle("alice")
        alice.processPreKeyBundle(bobBundle)

        val a = alice.encrypt("bob", "ping".toByteArray())
        assertEquals("ping", String(bob.decrypt("alice", a)))

        val b = bob.encrypt("alice", "pong".toByteArray())
        assertEquals(SignalProtocol.MESSAGE_TYPE_NORMAL, b.messageType)
        assertEquals("pong", String(alice.decrypt("bob", b)))
    }

    @Test
    fun `X3DH five sequential messages ratchet forward`() {
        val alice = SignalProtocol()
        val bob = SignalProtocol()
        val bobBundle = bob.generatePreKeyBundle("bob")
        alice.generatePreKeyBundle("alice")
        alice.processPreKeyBundle(bobBundle)

        for (i in 0 until 5) {
            val msg = byteArrayOf(i.toByte())
            val enc = alice.encrypt("bob", msg)
            assertEquals(i, enc.counter)
            val dec = bob.decrypt("alice", enc)
            assertContentEqualsKt(msg, dec)
        }
    }

    @Test
    fun `one-time pre-key consumed after responder establishes`() {
        val alice = SignalProtocol()
        val bob = SignalProtocol()
        val bobBundle = bob.generatePreKeyBundle("bob")
        alice.generatePreKeyBundle("alice")
        alice.processPreKeyBundle(bobBundle)

        val first = alice.encrypt("bob", "first".toByteArray())
        bob.decrypt("alice", first)

        // Replay using the same bundle should fail.
        val alice2 = SignalProtocol()
        alice2.generatePreKeyBundle("alice2")
        alice2.processPreKeyBundle(bobBundle)
        val replay = alice2.encrypt("bob", "replay".toByteArray())

        assertThrows<Exception> { bob.decrypt("alice2", replay) }
    }

    @Test
    fun `encrypt without local UHID throws`() {
        val alice = SignalProtocol()
        val bob = SignalProtocol()
        val bobBundle = bob.generatePreKeyBundle("bob")
        // Note: no generatePreKeyBundle / setLocalUhid on Alice.
        alice.processPreKeyBundle(bobBundle)
        assertThrows<IllegalStateException> { alice.encrypt("bob", "x".toByteArray()) }
    }

    @Test
    fun `pre-key bundle has both Ed25519 and X25519 identity keys`() {
        val svc = SignalProtocol()
        val bundle = svc.generatePreKeyBundle("alice")
        assertEquals(32, bundle.identityKey.size)         // Ed25519
        assertEquals(32, bundle.identityKeyX25519.size)   // X25519
        assertNotEquals(hex(bundle.identityKey), hex(bundle.identityKeyX25519))
        assertEquals(32, bundle.signedPreKey.size)
        assertEquals(32, bundle.preKey.size)
        assertEquals(64, bundle.signedPreKeySignature.size)
    }

    // ─── Double Ratchet (Signal §5) tests ──────────────────────────────────

    @Test
    fun `DoubleRatchet every message carries sender ephemeral key`() {
        val alice = SignalProtocol()
        val bob = SignalProtocol()
        val bobBundle = bob.generatePreKeyBundle("bob")
        alice.generatePreKeyBundle("alice")
        alice.processPreKeyBundle(bobBundle)

        val first = alice.encrypt("bob", "a".toByteArray())
        assertNotNull(first.senderEphemeralKeyX25519)
        assertEquals(32, first.senderEphemeralKeyX25519!!.size)

        bob.decrypt("alice", first)

        // Subsequent message also carries senderEphemeralKeyX25519 (same value
        // — Alice hasn't ratcheted because Bob hasn't responded yet).
        val second = alice.encrypt("bob", "b".toByteArray())
        assertNotNull(second.senderEphemeralKeyX25519)
        assertEquals(hex(first.senderEphemeralKeyX25519!!), hex(second.senderEphemeralKeyX25519!!))
    }

    @Test
    fun `DoubleRatchet sender ephemeral key rotates after roundtrip`() {
        val alice = SignalProtocol()
        val bob = SignalProtocol()
        val bobBundle = bob.generatePreKeyBundle("bob")
        alice.generatePreKeyBundle("alice")
        alice.processPreKeyBundle(bobBundle)

        // Alice -> Bob: Alice's first ratchet pub.
        val aliceFirst = alice.encrypt("bob", "ping".toByteArray())
        bob.decrypt("alice", aliceFirst)

        // Bob -> Alice: Bob's first ratchet pub (rotated by responder DH ratchet).
        val bobReply = bob.encrypt("alice", "pong".toByteArray())
        assertNotNull(bobReply.senderEphemeralKeyX25519)
        // Bob's ratchet pub MUST differ from Alice's (Bob rotated DHs on his
        // DH-ratchet step).
        assertNotEquals(hex(aliceFirst.senderEphemeralKeyX25519!!), hex(bobReply.senderEphemeralKeyX25519!!))

        alice.decrypt("bob", bobReply)

        // Alice -> Bob (after roundtrip): Alice rotates DHs on her own
        // DH-ratchet step (when she received Bob's reply).
        val aliceSecond = alice.encrypt("bob", "ping2".toByteArray())
        assertNotEquals(hex(aliceFirst.senderEphemeralKeyX25519!!), hex(aliceSecond.senderEphemeralKeyX25519!!))
        assertNotEquals(hex(bobReply.senderEphemeralKeyX25519!!), hex(aliceSecond.senderEphemeralKeyX25519!!))

        // Bob can still decrypt Alice's new message.
        val dec = bob.decrypt("alice", aliceSecond)
        assertEquals("ping2", String(dec))
    }

    @Test
    fun `DoubleRatchet previous chain count tracks messages per chain`() {
        val alice = SignalProtocol()
        val bob = SignalProtocol()
        val bobBundle = bob.generatePreKeyBundle("bob")
        alice.generatePreKeyBundle("alice")
        alice.processPreKeyBundle(bobBundle)

        // Alice sends 3 messages without a roundtrip.
        for (i in 0 until 3) {
            val enc = alice.encrypt("bob", "a$i".toByteArray())
            // PN is 0 because this IS Alice's first chain.
            assertEquals(0, enc.previousChainCount)
            bob.decrypt("alice", enc)
        }

        // Bob sends a reply, triggering his DH-ratchet step.
        val bobReply = bob.encrypt("alice", "hi".toByteArray())
        // Bob's PN reflects however many messages Bob sent in his previous
        // sending chain — which was 0 (Bob hadn't sent before his ratchet
        // rotated his chain).
        assertEquals(0, bobReply.previousChainCount)
        alice.decrypt("bob", bobReply)

        // Alice's next message after her DH-ratchet step. PN should be 3 —
        // that's how many messages she sent on her previous chain before
        // Bob's reply triggered her ratchet.
        val aliceNew = alice.encrypt("bob", "a3".toByteArray())
        assertEquals(3, aliceNew.previousChainCount)
    }

    @Test
    fun `DoubleRatchet out-of-order across DH-ratchet boundary still decrypts`() {
        // Alice sends 3 messages on chain 1. Bob receives only the first 2,
        // then Alice does a DH-ratchet (Bob replied) and sends a 4th on chain
        // 2. The 3rd message (from chain 1) arrives last — Bob must decrypt
        // via the skipped-keys cache keyed by (Alice's old DHs pub, counter=2).
        val alice = SignalProtocol()
        val bob = SignalProtocol()
        val bobBundle = bob.generatePreKeyBundle("bob")
        alice.generatePreKeyBundle("alice")
        alice.processPreKeyBundle(bobBundle)

        val a0 = alice.encrypt("bob", "a0".toByteArray())
        val a1 = alice.encrypt("bob", "a1".toByteArray())
        val a2 = alice.encrypt("bob", "a2".toByteArray())

        // Bob receives a0, a1 only.
        assertEquals("a0", String(bob.decrypt("alice", a0)))
        assertEquals("a1", String(bob.decrypt("alice", a1)))

        // Bob replies — triggers his DH-ratchet step.
        val bReply = bob.encrypt("alice", "hi".toByteArray())
        alice.decrypt("bob", bReply)

        // Alice sends a4 on her new chain (after her DH-ratchet step).
        val a4 = alice.encrypt("bob", "a4".toByteArray())
        // Bob receives a4 — triggers his second DH-ratchet step. He must
        // skip-derive a key for Alice's old chain counter=2 because PN=3.
        assertEquals("a4", String(bob.decrypt("alice", a4)))

        // Now the missing a2 (from Alice's OLD chain) finally arrives. Bob
        // pulls the skipped key from cache.
        assertEquals("a2", String(bob.decrypt("alice", a2)))
    }

    @Test
    fun `DoubleRatchet long conversation all messages decrypt`() {
        val alice = SignalProtocol()
        val bob = SignalProtocol()
        val bobBundle = bob.generatePreKeyBundle("bob")
        alice.generatePreKeyBundle("alice")
        alice.processPreKeyBundle(bobBundle)

        // 10 alternating messages — each side ratchets at every roundtrip.
        for (i in 0 until 10) {
            val aMsg = "alice $i"
            val aEnc = alice.encrypt("bob", aMsg.toByteArray())
            assertEquals(aMsg, String(bob.decrypt("alice", aEnc)))

            val bMsg = "bob $i"
            val bEnc = bob.encrypt("alice", bMsg.toByteArray())
            assertEquals(bMsg, String(alice.decrypt("bob", bEnc)))
        }
    }

    // ─── Helpers ───────────────────────────────────────────────────────────

    private fun loadFixturePair(caseName: String): Pair<Map<String, String>, Map<String, String>> {
        val root = repoRoot()
        val inputsFile = File(root, "fixtures/signal/inputs.json")
        val expectedFile = File(root, "fixtures/signal/expected/${caseName}.json")
        val inputsCases = parseCases(inputsFile.readText())
        val inputs = inputsCases.first { it["name"] == caseName }
        val expected = parseFlatJson(expectedFile.readText())
        return inputs to expected
    }

    private fun repoRoot(): File {
        var dir: File? = File(".").canonicalFile
        repeat(10) {
            val candidate = File(dir, "AetherProtocol.slnx")
            if (candidate.exists()) return dir!!
            dir = dir?.parentFile ?: return@repeat
        }
        throw IllegalStateException("AetherProtocol.slnx not found from ${File(".").canonicalFile}")
    }

    /** Minimal JSON parser for our fixed-shape fixture files (string -> string maps; cases array). */
    private fun parseCases(json: String): List<Map<String, String>> {
        // Slice between "cases":[ and matching closing ]
        val casesIdx = json.indexOf("\"cases\"")
        require(casesIdx >= 0) { "No cases array in inputs.json" }
        val openBracket = json.indexOf('[', casesIdx)
        var depth = 0
        var endIdx = -1
        for (i in openBracket until json.length) {
            when (json[i]) {
                '[' -> depth++
                ']' -> { depth--; if (depth == 0) { endIdx = i; break } }
            }
        }
        require(endIdx > 0) { "Unbalanced cases array" }
        val arrayBody = json.substring(openBracket + 1, endIdx)
        return splitObjects(arrayBody).map { parseFlatJson(it) }
    }

    private fun splitObjects(text: String): List<String> {
        val result = mutableListOf<String>()
        var depth = 0
        var start = -1
        for (i in text.indices) {
            when (text[i]) {
                '{' -> { if (depth == 0) start = i; depth++ }
                '}' -> { depth--; if (depth == 0 && start >= 0) { result += text.substring(start, i + 1); start = -1 } }
            }
        }
        return result
    }

    private fun parseFlatJson(json: String): Map<String, String> {
        // Parses a JSON object with string-only values. Ignores numeric / structured fields.
        val result = mutableMapOf<String, String>()
        val regex = Regex("\"([A-Za-z_0-9\\$]+)\"\\s*:\\s*\"((?:[^\"\\\\]|\\\\.)*)\"")
        for (match in regex.findAll(json)) {
            result[match.groupValues[1]] = match.groupValues[2]
        }
        return result
    }

    private fun unhex(s: String): ByteArray {
        require(s.length % 2 == 0) { "hex string must have even length" }
        return ByteArray(s.length / 2) { i ->
            ((Character.digit(s[i * 2], 16) shl 4) + Character.digit(s[i * 2 + 1], 16)).toByte()
        }
    }

    private fun hex(b: ByteArray): String =
        b.joinToString("") { "%02x".format(it.toInt() and 0xFF) }

    private fun x25519DerivePub(priv: ByteArray): ByteArray {
        val p = X25519PrivateKeyParameters(priv, 0)
        return p.generatePublicKey().encoded
    }

    private fun x25519AgreeStandalone(priv: ByteArray, pub: ByteArray): ByteArray {
        val p = X25519PrivateKeyParameters(priv, 0)
        val q = X25519PublicKeyParameters(pub, 0)
        val agreement = X25519Agreement()
        agreement.init(p)
        val shared = ByteArray(agreement.agreementSize)
        agreement.calculateAgreement(q, shared, 0)
        return shared
    }

    private fun hkdf32Standalone(ikm: ByteArray, info: ByteArray): ByteArray {
        val salt = ByteArray(32)
        val mac1 = Mac.getInstance("HmacSHA256")
        mac1.init(SecretKeySpec(salt, "HmacSHA256"))
        val prk = mac1.doFinal(ikm)
        val mac2 = Mac.getInstance("HmacSHA256")
        mac2.init(SecretKeySpec(prk, "HmacSHA256"))
        mac2.update(info)
        mac2.update(0x01.toByte())
        return mac2.doFinal().copyOf(32)
    }

    private fun hmacOne(key: ByteArray, b: Byte): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(byteArrayOf(b))
    }

    private fun assertContentEqualsKt(expected: ByteArray, actual: ByteArray) {
        if (!expected.contentEquals(actual)) {
            kotlin.test.fail("byte arrays differ: expected ${hex(expected)}, got ${hex(actual)}")
        }
    }
}

File: go/security/signal_fixture_test.go
// SPDX-License-Identifier: MIT

package security

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"golang.org/x/crypto/curve25519"
	"golang.org/x/crypto/hkdf"
)

// Cross-language Signal-protocol fixture verifier. Runs the X3DH and HMAC
// ratchet steps against the canonical inputs in fixtures/signal/inputs.json
// and asserts the outputs match fixtures/signal/expected/*.json byte-for-byte.
//
// Any drift between this Go implementation and the C# reference (or any
// other language) shows up here as a hex mismatch.

type signalInputs struct {
	Cases []map[string]any `json:"cases"`
}

func TestSignalFixture_X3DHBasic(t *testing.T) {
	inputs, expected := loadFixturePair(t, "x3dh_basic")

	aliceIK := mustHex(t, inputs["alice_identity_priv_hex"].(string))
	aliceEK := mustHex(t, inputs["alice_ephemeral_priv_hex"].(string))
	bobIK := mustHex(t, inputs["bob_identity_priv_hex"].(string))
	bobSPK := mustHex(t, inputs["bob_signed_pre_key_priv_hex"].(string))
	bobOPK := mustHex(t, inputs["bob_one_time_pre_key_priv_hex"].(string))

	aliceIKPub := mustDerive(t, aliceIK)
	aliceEKPub := mustDerive(t, aliceEK)
	bobIKPub := mustDerive(t, bobIK)
	bobSPKPub := mustDerive(t, bobSPK)
	bobOPKPub := mustDerive(t, bobOPK)

	// Initiator-side X3DH.
	dh1, err := curve25519.X25519(aliceIK, bobSPKPub)
	if err != nil {
		t.Fatalf("DH1: %v", err)
	}
	dh2, err := curve25519.X25519(aliceEK, bobIKPub)
	if err != nil {
		t.Fatalf("DH2: %v", err)
	}
	dh3, err := curve25519.X25519(aliceEK, bobSPKPub)
	if err != nil {
		t.Fatalf("DH3: %v", err)
	}
	dh4, err := curve25519.X25519(aliceEK, bobOPKPub)
	if err != nil {
		t.Fatalf("DH4: %v", err)
	}

	shared := append(append(append(append([]byte{}, dh1...), dh2...), dh3...), dh4...)
	rootInfo := []byte(inputs["hkdf_root_info_utf8"].(string))
	sendInfo := []byte(inputs["hkdf_chain_initiator_send_info_utf8"].(string))
	recvInfo := []byte(inputs["hkdf_chain_initiator_recv_info_utf8"].(string))

	rootKey := hkdfDerive(shared, rootInfo)
	sendChain := hkdfDerive(rootKey, sendInfo)
	recvChain := hkdfDerive(rootKey, recvInfo)

	checks := map[string][]byte{
		"alice_identity_pub_hex":       aliceIKPub,
		"alice_ephemeral_pub_hex":      aliceEKPub,
		"bob_identity_pub_hex":         bobIKPub,
		"bob_signed_pre_key_pub_hex":   bobSPKPub,
		"bob_one_time_pre_key_pub_hex": bobOPKPub,
		"dh1_hex":                      dh1,
		"dh2_hex":                      dh2,
		"dh3_hex":                      dh3,
		"dh4_hex":                      dh4,
		"shared_secret_hex":            shared,
		"root_key_hex":                 rootKey,
		"initiator_send_chain_key_hex": sendChain,
		"initiator_recv_chain_key_hex": recvChain,
	}
	verifyAll(t, "x3dh_basic", expected, checks)
}

func TestSignalFixture_RatchetStepBasic(t *testing.T) {
	inputs, expected := loadFixturePair(t, "ratchet_step_basic")
	chainKey := mustHex(t, inputs["chain_key_hex"].(string))

	verifyAll(t, "ratchet_step_basic", expected, map[string][]byte{
		"message_key_hex":    hmacOne(chainKey, 0x01),
		"next_chain_key_hex": hmacOne(chainKey, 0x02),
	})
}

func TestSignalFixture_RatchetStepThreeIterations(t *testing.T) {
	inputs, expected := loadFixturePair(t, "ratchet_step_three_iterations")
	chainKey := mustHex(t, inputs["initial_chain_key_hex"].(string))

	checks := make(map[string][]byte)
	for i := 0; i < 3; i++ {
		msgKey := hmacOne(chainKey, 0x01)
		next := hmacOne(chainKey, 0x02)
		checks[stepKey(i, "message_key_hex")] = msgKey
		checks[stepKey(i, "chain_key_after_hex")] = next
		chainKey = next
	}
	verifyAll(t, "ratchet_step_three_iterations", expected, checks)
}

// ─── End-to-end exercises of the X3DH + ratchet flow ─────────────────────

func TestX3DH_EndToEnd_FirstMessageRoundTrips(t *testing.T) {
	alice, _ := NewSignalProtocolService()
	bob, _ := NewSignalProtocolService()

	bobBundle, err := bob.GeneratePreKeyBundle("bob")
	if err != nil {
		t.Fatalf("Bob.GeneratePreKeyBundle: %v", err)
	}
	if _, err := alice.GeneratePreKeyBundle("alice"); err != nil {
		t.Fatalf("Alice.GeneratePreKeyBundle: %v", err)
	}
	if err := alice.ProcessPreKeyBundle(bobBundle); err != nil {
		t.Fatalf("Alice.ProcessPreKeyBundle: %v", err)
	}

	encrypted, err := alice.Encrypt("bob", []byte("the mesh is alive"))
	if err != nil {
		t.Fatalf("Alice.Encrypt: %v", err)
	}
	if encrypted.MessageType != MessageTypePreKey {
		t.Fatalf("first msg should be PreKey (1), got %d", encrypted.MessageType)
	}
	if len(encrypted.InitiatorIdentityKeyX25519) != 32 || len(encrypted.InitiatorEphemeralKeyX25519) != 32 {
		t.Fatalf("PreKey msg missing IK/EK")
	}
	if encrypted.SenderUhid != "alice" {
		t.Fatalf("SenderUhid = %q, want alice", encrypted.SenderUhid)
	}

	plaintext, err := bob.Decrypt("alice", encrypted)
	if err != nil {
		t.Fatalf("Bob.Decrypt: %v", err)
	}
	if string(plaintext) != "the mesh is alive" {
		t.Fatalf("plaintext mismatch: %q", string(plaintext))
	}
}

func TestX3DH_SubsequentMessage_IsNormalNotPreKey(t *testing.T) {
	alice, _ := NewSignalProtocolService()
	bob, _ := NewSignalProtocolService()

	bobBundle, _ := bob.GeneratePreKeyBundle("bob")
	alice.GeneratePreKeyBundle("alice")
	alice.ProcessPreKeyBundle(bobBundle)

	first, _ := alice.Encrypt("bob", []byte("a"))
	bob.Decrypt("alice", first)

	second, err := alice.Encrypt("bob", []byte("b"))
	if err != nil {
		t.Fatalf("second encrypt: %v", err)
	}
	if second.MessageType != MessageTypeNormal {
		t.Fatalf("second msg should be normal (0), got %d", second.MessageType)
	}
	if len(second.InitiatorIdentityKeyX25519) != 0 {
		t.Fatalf("second msg should not carry IK")
	}

	out, err := bob.Decrypt("alice", second)
	if err != nil {
		t.Fatalf("Bob.Decrypt second: %v", err)
	}
	if string(out) != "b" {
		t.Fatalf("plaintext = %q, want b", string(out))
	}
}

func TestX3DH_BidirectionalAfterFirstMessage(t *testing.T) {
	alice, _ := NewSignalProtocolService()
	bob, _ := NewSignalProtocolService()

	bobBundle, _ := bob.GeneratePreKeyBundle("bob")
	alice.GeneratePreKeyBundle("alice")
	alice.ProcessPreKeyBundle(bobBundle)

	a, _ := alice.Encrypt("bob", []byte("ping"))
	if _, err := bob.Decrypt("alice", a); err != nil {
		t.Fatalf("Bob.Decrypt ping: %v", err)
	}

	b, err := bob.Encrypt("alice", []byte("pong"))
	if err != nil {
		t.Fatalf("Bob.Encrypt: %v", err)
	}
	if b.MessageType != MessageTypeNormal {
		t.Fatalf("Bob's reply should be normal (0), got %d", b.MessageType)
	}
	out, err := alice.Decrypt("bob", b)
	if err != nil {
		t.Fatalf("Alice.Decrypt pong: %v", err)
	}
	if string(out) != "pong" {
		t.Fatalf("plaintext = %q, want pong", string(out))
	}
}

func TestX3DH_FiveSequentialMessages_RatchetsForward(t *testing.T) {
	alice, _ := NewSignalProtocolService()
	bob, _ := NewSignalProtocolService()

	bobBundle, _ := bob.GeneratePreKeyBundle("bob")
	alice.GeneratePreKeyBundle("alice")
	alice.ProcessPreKeyBundle(bobBundle)

	for i := 0; i < 5; i++ {
		msg := []byte{byte(i)}
		enc, err := alice.Encrypt("bob", msg)
		if err != nil {
			t.Fatalf("encrypt %d: %v", i, err)
		}
		if enc.Counter != int32(i) {
			t.Fatalf("counter %d, want %d", enc.Counter, i)
		}
		dec, err := bob.Decrypt("alice", enc)
		if err != nil {
			t.Fatalf("decrypt %d: %v", i, err)
		}
		if len(dec) != 1 || dec[0] != byte(i) {
			t.Fatalf("plaintext %d mismatch", i)
		}
	}
}

func TestX3DH_OneTimePreKey_ConsumedAfterResponderEstablishes(t *testing.T) {
	alice, _ := NewSignalProtocolService()
	bob, _ := NewSignalProtocolService()

	bobBundle, _ := bob.GeneratePreKeyBundle("bob")
	alice.GeneratePreKeyBundle("alice")
	alice.ProcessPreKeyBundle(bobBundle)

	first, _ := alice.Encrypt("bob", []byte("first"))
	if _, err := bob.Decrypt("alice", first); err != nil {
		t.Fatalf("Bob.Decrypt first: %v", err)
	}

	// A second initiator using the same bundle (and therefore same OPK id)
	// should fail because Bob consumed the OPK.
	alice2, _ := NewSignalProtocolService()
	alice2.GeneratePreKeyBundle("alice2")
	alice2.ProcessPreKeyBundle(bobBundle)
	replay, _ := alice2.Encrypt("bob", []byte("replay"))

	if _, err := bob.Decrypt("alice2", replay); err == nil {
		t.Fatalf("expected error decrypting replayed OPK msg, got nil")
	}
}

func TestEncrypt_WithoutLocalUhid_ReturnsError(t *testing.T) {
	alice, _ := NewSignalProtocolService()
	bob, _ := NewSignalProtocolService()
	bobBundle, _ := bob.GeneratePreKeyBundle("bob")
	// Note: no GeneratePreKeyBundle / SetLocalUhid on Alice.
	if err := alice.ProcessPreKeyBundle(bobBundle); err != nil {
		t.Fatalf("ProcessPreKeyBundle: %v", err)
	}
	if _, err := alice.Encrypt("bob", []byte("x")); err == nil {
		t.Fatalf("expected error when local UHID unset")
	}
}

func TestPreKeyBundle_HasBothIdentityKeys(t *testing.T) {
	svc, _ := NewSignalProtocolService()
	bundle, err := svc.GeneratePreKeyBundle("alice")
	if err != nil {
		t.Fatalf("GeneratePreKeyBundle: %v", err)
	}
	if len(bundle.IdentityKey) != 32 {
		t.Fatalf("Ed25519 identity key length = %d, want 32", len(bundle.IdentityKey))
	}
	if len(bundle.IdentityKeyX25519) != 32 {
		t.Fatalf("X25519 identity key length = %d, want 32", len(bundle.IdentityKeyX25519))
	}
	if len(bundle.SignedPreKey) != 32 {
		t.Fatalf("SPK length = %d, want 32", len(bundle.SignedPreKey))
	}
	if len(bundle.PreKey) != 32 {
		t.Fatalf("OPK length = %d, want 32", len(bundle.PreKey))
	}
	if len(bundle.SignedPreKeySignature) != 64 {
		t.Fatalf("SPK signature length = %d, want 64", len(bundle.SignedPreKeySignature))
	}
}

// ─── Helpers ─────────────────────────────────────────────────────────────

func loadFixturePair(t *testing.T, caseName string) (inputs, expected map[string]any) {
	t.Helper()
	root := repoRoot(t)
	inputsPath := filepath.Join(root, "fixtures", "signal", "inputs.json")
	expectedPath := filepath.Join(root, "fixtures", "signal", "expected", caseName+".json")

	inputsBytes, err := os.ReadFile(inputsPath)
	if err != nil {
		t.Fatalf("read inputs: %v", err)
	}
	var doc signalInputs
	if err := json.Unmarshal(inputsBytes, &doc); err != nil {
		t.Fatalf("unmarshal inputs: %v", err)
	}
	for _, c := range doc.Cases {
		if c["name"] == caseName {
			inputs = c
			break
		}
	}
	if inputs == nil {
		t.Fatalf("case %q not in inputs.json", caseName)
	}

	expectedBytes, err := os.ReadFile(expectedPath)
	if err != nil {
		t.Fatalf("read expected: %v", err)
	}
	if err := json.Unmarshal(expectedBytes, &expected); err != nil {
		t.Fatalf("unmarshal expected: %v", err)
	}
	return inputs, expected
}

func verifyAll(t *testing.T, caseName string, expected map[string]any, actuals map[string][]byte) {
	t.Helper()
	for k, v := range actuals {
		exp, ok := expected[k].(string)
		if !ok {
			t.Errorf("[%s] expected fixture missing field %q", caseName, k)
			continue
		}
		got := hex.EncodeToString(v)
		if exp != got {
			t.Errorf("[%s] %s mismatch:\n  expected: %s\n  actual:   %s", caseName, k, exp, got)
		}
	}
}

func repoRoot(t *testing.T) string {
	t.Helper()
	dir, err := os.Getwd()
	if err != nil {
		t.Fatalf("getwd: %v", err)
	}
	for {
		if _, err := os.Stat(filepath.Join(dir, "AetherProtocol.slnx")); err == nil {
			return dir
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			t.Fatalf("repo root not found from %s", dir)
		}
		dir = parent
	}
}

func mustHex(t *testing.T, s string) []byte {
	t.Helper()
	b, err := hex.DecodeString(s)
	if err != nil {
		t.Fatalf("hex.DecodeString: %v", err)
	}
	return b
}

func mustDerive(t *testing.T, priv []byte) []byte {
	t.Helper()
	pub, err := curve25519.X25519(priv, curve25519.Basepoint)
	if err != nil {
		t.Fatalf("X25519 base: %v", err)
	}
	return pub
}

func hkdfDerive(ikm, info []byte) []byte {
	h := hkdf.New(sha256.New, ikm, nil, info)
	out := make([]byte, 32)
	h.Read(out)
	return out
}

func hmacOne(key []byte, b byte) []byte {
	h := hmac.New(sha256.New, key)
	h.Write([]byte{b})
	return h.Sum(nil)
}

func stepKey(i int, suffix string) string {
	return "step_" + string(rune('0'+i)) + "_" + suffix
}
